### PR TITLE
Create ProposalSalt event for TimelockController

### LINKED ISSUE #3216: TimelockController 'salt' is unrecoverable yet required for execution
When scheduling a transaction using the `TimelockController`, one is required to provide a `salt` to distinguish otherwise identical transactions. The `salt` parameter is not emitted in the `CallScheduled` event which makes it impossible to recover. Yet `salt` is required for executing the scheduled transaction transaction, ~~even though it isn't necessary and more gas efficient if `id` is used instead~~.

**💻 Environment**

Openzeppelin v4.5.0  / hardhat

**📝 Details**

`salt` is missing from the `CallScheduled` event: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/458697be32f41946e1ab66e946c7cf373c79577d/contracts/governance/TimelockController.sol#L35

`salt` is required for the `execute` & `executeBatch` calls: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/458697be32f41946e1ab66e946c7cf373c79577d/contracts/governance/TimelockController.sol#L267

~~Instead of `salt` the `execute` methods should be consuming the hashed operation (`id`) directly, since that id is readily available via event data. This would also save on gas since there is no need to re-run the hashing operation.~~

Happy to submit a PR

### PR DESCRIPTION
**Pull request by JulissaDantes**
Fixes #3216  
Fixes LIB-639

When scheduling a transaction using the TimelockController, one is required to provide a salt to distinguish otherwise identical transactions. The salt parameter is not emitted in the CallScheduled event which makes it impossible to recover. This PR introduces the `ProposalSalt` event to get the salt after scheduling an operation, if salt == 0 it will not emit an event.

#### PR Checklist

- [X] Tests
- [X] Documentation
- [X] Changeset entry (run `npx changeset add`)

### CODE REVIEW

File: contracts/governance/TimelockController.sol [hunk 1]
```diff
@@ -249,6 +257,9 @@ contract TimelockController is AccessControl, IERC721Receiver, IERC1155Receiver
         _schedule(id, delay);
         for (uint256 i = 0; i < targets.length; ++i) {
             emit CallScheduled(id, i, targets[i], values[i], payloads[i], predecessor, delay);
+            if (salt != bytes32(0)) {
```
Comment (Amxx) [inline contracts/governance/TimelockController.sol:260]: This should not be inside the loop.

File: contracts/governance/TimelockController.sol [hunk 2]
```diff
@@ -51,6 +51,11 @@ contract TimelockController is AccessControl, IERC721Receiver, IERC1155Receiver
      */
     event CallExecuted(bytes32 indexed id, uint256 indexed index, address target, uint256 value, bytes data);
 
+    /**
+     * @dev Emitted when new proposal is scheduled.
+     */
+    event ProposalSalt(bytes32 indexed proposalId, bytes32 salt);
```
Comment (Amxx) [inline contracts/governance/TimelockController.sol:57]: I think we should use `id` for naming consistency accross events

File: .changeset/five-poets-mix.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'openzeppelin-solidity': patch
+---
+
+`TimelockController`: Add the `ProposalSalt` event to emit on operation schedule.([#4001](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4001))
```
Comment (frangio) [inline .changeset/five-poets-mix.md:5]:
```suggestion
`TimelockController`: Add the `ProposalSalt` event to emit on operation schedule.
```

File: contracts/governance/TimelockController.sol [hunk 4]
```diff
@@ -223,6 +228,9 @@ contract TimelockController is AccessControl, IERC721Receiver, IERC1155Receiver
         bytes32 id = hashOperation(target, value, data, predecessor, salt);
         _schedule(id, delay);
         emit CallScheduled(id, 0, target, value, data, predecessor, delay);
+        if (salt != bytes32(0)) {
+            emit ProposalSalt(id, salt);
```
Comment (frangio) [inline contracts/governance/TimelockController.sol:232]: The functoin documentation says that `CallScheduled` is emitted. We need to add `ProposalSalt` in there as well. Same in `scheduleBatch`.

File: contracts/governance/TimelockController.sol [hunk 5]
```diff
@@ -51,6 +51,11 @@ contract TimelockController is AccessControl, IERC721Receiver, IERC1155Receiver
      */
     event CallExecuted(bytes32 indexed id, uint256 indexed index, address target, uint256 value, bytes data);
 
+    /**
+     * @dev Emitted when new proposal is scheduled.
+     */
+    event ProposalSalt(bytes32 indexed id, bytes32 salt);
```
Comment (frangio) [inline contracts/governance/TimelockController.sol:57]: The issue suggested `ProposalSalt` but the rest of the events are `CallExecuted`, `CallScheduled`, etc.

This should probably be called `CallSalt`. @Amxx?

File: contracts/governance/TimelockController.sol [hunk 6]
```diff
@@ -51,6 +51,11 @@ contract TimelockController is AccessControl, IERC721Receiver, IERC1155Receiver
      */
     event CallExecuted(bytes32 indexed id, uint256 indexed index, address target, uint256 value, bytes data);
 
+    /**
+     * @dev Emitted when new proposal is scheduled.
```
Comment (frangio) [inline contracts/governance/TimelockController.sol:55]: We need to expain it's only emitted for non-zero salt.

**Review by frangio — CHANGES_REQUESTED** [review]

File: contracts/governance/TimelockController.sol (hunk 5 above)
Comment (Amxx) [reply contracts/governance/TimelockController.sol:57]: Yes, `CallSalt` would make more sens.

**Review by frangio — APPROVED** [review]
Thanks!

**Pull request merged**